metrotranscom/doorway
# fix: switch bulk emails to individual email sends

This PR addresses #986 #983 #985 #1008


## Description

The current setup for emails with multiple recipients has the following problems:
- Unable to send to more than 50 emails at a time (limit from AWS)
- Unable to track if a user got an email if there is more than one recipient of the email
- Limited logging makes it hard to see if group emails were sent via the backend

This change resolves the above issues by doing the following:
- Switch to using the v2 version of the AWS SES sdk and no longer use nodemailer
- All emails are now individually sent  and recipient is in the `to` rather than `bcc`
- Use the SendBulkEmailCommand which can bulk send 50 emails at a time
  - All bulk emails are now chunked into groups of 50 emails and for each group one bulk send is sent
- Add logger calls before every bulk email send with the number of emails it is sending to
- Add additional logging for failed email sends
- Lottery emails are also sent to an environment variable designated email address

Along with the above the following bugs/issues were addressed:
- New listing notification email was missing several community types
- The applicant lottery published email had the placeholder `exygy.com` url instead of the correct ones for the notificationsUrl and helpCenterUrl
- All user related emails did not have an await. This was not causing an issue, but if a failure happened it would have been harder to track since the user potentially could have got the response before email actually sent
- Update tests to make sure the correct emails are sent

## How Can This Be Tested/Reviewed?

Since this change touches how all emails are sent in Doorway we will need to test all of the different emails sent out

Here is documentation for the the list of emails https://docs.google.com/document/d/1bPl-SVapMcktRt06u-vaT1Lial4OR-gpe__vU0Dgt4c/edit?tab=t.0 

To test the site email is also sent during the three lottery emails you will need to add a value to the api's .env file. After adding this you should receive an email for lotteryReleased, lotteryPublishedAdmin, and lotteryPublishedApplicant email sends
Sample:
`SITE_EMAIL=morgan.ludtke+siteemail@exygy.com`

For the bulk emails some additional testing should be done.
Scenarios:
- Trigger an email that will go out to more than 50 people
An easy way test many email send you could add this code to the seed-staging.ts file before line 349 (change the email to your email)
```
  const bulkApplications = [];
  for (let i = 0; i <= 51; i++) {
    const user = await prismaClient.userAccounts.create({
      data: await userFactory({
        email: `morgan.ludtke+${i}@exygy.com`,
        confirmedAt: new Date(),
        jurisdictionIds: [jurisdiction.id],
        acceptedTerms: true,
        password: 'abcdef',
      }),
    });
    bulkApplications.push(
      await applicationFactory({
        applicant: { emailAddress: `morgan.ludtke+${i}@exygy.com` },
        userId: user.id,
      }),
    );
  }
```
Then change the applications list to that bulkApplications list for any of the listings
```
applications: bulkApplications
```
Then you should be able to do the whole lottery release flow to receive 50+ application emails

## Author Checklist:


## Review Process:

- Read and understand the issue
- Ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Either (1) explicitly ask a clarifying question, (2) request changes, or (3) approve the PR, even if there are very small remaining changes, if you don't need to re-review after the updates
